### ISSUE TITLE
Add Support for Device/Distribution/Collection Metadata Endpoints

### ISSUE THREAD

**Issue by kristinpeterson** [attm2x/m2x-nanode#4]
See [Ruby library implementation](https://github.com/attm2x/m2x-ruby/commit/044b282afd614e87e68b7516353d65cbe3c79ab0) for reference.

Add support for the following endpoints:

Device Metadata:
* https://m2x.att.com/developer/documentation/v2/device#Read-Device-Metadata
* https://m2x.att.com/developer/documentation/v2/device#Update-Device-Metadata
* https://m2x.att.com/developer/documentation/v2/device#Read-Device-Metadata-Field
* https://m2x.att.com/developer/documentation/v2/device#Update-Device-Metadata-Field

Distribution Metadata:
* https://m2x.att.com/developer/documentation/v2/distribution#Read-Distribution-Metadata
* https://m2x.att.com/developer/documentation/v2/distribution#Update-Distribution-Metadata
* https://m2x.att.com/developer/documentation/v2/distribution#Read-Distribution-Metadata-Field
* https://m2x.att.com/developer/documentation/v2/distribution#Update-Distribution-Metadata-Field

Collection Metadata:
* https://m2x.att.com/developer/documentation/v2/collections#Read-Collection-Metadata
* https://m2x.att.com/developer/documentation/v2/collections#Update-Collection-Metadata
* https://m2x.att.com/developer/documentation/v2/collections#Read-Collection-Metadata-Field
* https://m2x.att.com/developer/documentation/v2/collections#Update-Collection-Metadata-Field

**Comment by xxuejie** [comment]
I will just use this single issue to raise a discussion, while the same applies to more than one issue in a couple of libraries(including Arduino, ARM Mbed, Nanode, Cypress PSoC, etc), I prefer not to mention that anywhere so as not to spam everyone's inbox.

My point is: I don't think it will be useful to add metadata/command/job APIs to the libraries used by embedded boards. If you check out the library involved here, most of them only have a handful of APIs for writing data, some of them(such as Nanode) don't even support reading data due to certain limitation. So it remains a question if we can implement those APIs on those boards, and even if we can, I don't think there will be any practical uses.

The only API which will be useful here, in my opinion, is [Post Device Update](https://m2x.att.com/developer/documentation/v2/device#Post-Device-Update--Single-Values-to-Multiple-Streams-), and we can implement it across those embedded boards. But I don't feel we need to implement the rest APIs.

Thoughts? @kristinpeterson

**Comment by fsaravia** [comment]
I agree with @xxuejie, I think we can leave metadata to just the full featured devices.

Regarding commands (which is beyond the scope of this issue but was mentioned above), these libraries should only implement the device's side of the API: requesting/accepting/rejecting commands.

And regarding Jobs API, I also agree, it makes sense to add those only on libraries with search/read values and is not the case for this kind of libraries, so I would leave it out as well.

@kristinpeterson: thoughts? we can add this to the guidelines if you agree

**Comment by xxuejie** [comment]
@fsaravia I know some APIs do not belong here, but I feel using a single place for discussion might be better (or not :P).

After a second thought, I agree with @fsaravia that commands can be helpful here, and I'm going to implement that in those affected libraries.

However, one potential problem we might have, is that certain libraries (such as Nanode) might not have enough capabilities for parsing JSON, hence we cannot implement the API to request pending commands, which would render accepting/rejecting commands useless. Those boards would certain need help from the server to mitigate the problem, we will need some thoughts before deciding which is the best way to solve this problem.

**Comment by kristinpeterson** [comment]
I agree regarding leaving out the Jobs API, listing Commands & reading Metadata for limited libraries. However I do have the following comments:

**Processing / Rejecting Command**

> However, one potential problem we might have, is that certain libraries (such as Nanode) might not have enough capabilities for parsing JSON, hence we cannot implement the API to request pending commands, which would render accepting/rejecting commands useless.

I think that a limited device should still be able to [process](https://m2x.att.com/developer/documentation/v2/commands#Device-Marks-a-Command-as-Processed)/[reject](https://m2x.att.com/developer/documentation/v2/commands#Device-Marks-a-Command-as-Rejected) a command, to allow the device to update the command's status. @xxuejie any reason why the Nanode, or any other limited device we currently support would not be able to do so?

**Updating Metadata**

This might be a valuable feature, but for now - let's err on the side of not implementing any of the Metadata endpoints for limited libraries. If this feature is requested by a user in the future, we can reconsider this choice.

---

TL;DR

I propose we add [process](https://m2x.att.com/developer/documentation/v2/commands#Device-Marks-a-Command-as-Processed) & [reject](https://m2x.att.com/developer/documentation/v2/commands#Device-Marks-a-Command-as-Rejected) Command endpoint support for Limited Libraries, and leave out Jobs API, all Metadata endpoints & listing/reading commands endpoints.

I'll update the [Supported API methods](https://github.com/attm2x/m2x-service/blob/master/LIMITED-CLIENT-CONTRIBUTIONS.md#supported-api-methods) section of the Limited Client Contribution Guidelines, if we all agree.

**Comment by jemc** [comment]
@xxuejie does MessagePack (or BSON?) make this any more feasible than JSON?

**Comment by xxuejie** [comment]
@kristinpeterson The reason I'm leaving out Nanode, is that we cannot implement the API to pull down current received commands. If we cannot fetch a command to process, I wonder if process/reject is gonna be useful here.

@jemc I will really have to test it, but I must say I'm not being optimistic on this one here.
